README
easyGestures N is a Firefox extension that provides a pie menu for browsing the Web from where your cursor is currently located.

It is a fork of the easyGestures extension created by Ons Besbes, who maintained it until Firefox 4. This extension worked fine with subsequent versions of Firefox, although it required changing the maxVersion value in its install manifest. However, with Firefox 14 easyGestures started behaving erratically and generating "TypeError: can't access dead object" errors.

easyGestures N is an (ongoing) attempt to bring easyGestures back to a usable and more maintainable state. This includes pruning those features that IMO fall outside the core concept of the extension. Help is very much welcome, especially for updating and maintaining both the visual appearance and the current locales.

To install the latest development version:
- Uninstall any version of easyGestures prior to version 4.3.2
- Click on the button labeled "Download ZIP"
- Once the file downloaded, unzip it
- Zip the content of the "easyGestures-development" folder (and not the folder itself!)
- Give to the resulting zip file the "xpi" extension
- Drop the renamed zipped file into Firefox

CHANGELOG
Changes since version 4.3.2 of easyGestures by Ons Besbes (see git history for the full details: https://github.com/ngdeleito/easyGestures/commits/development)


Version 4.9 (March 15, 2015)

- revised and updated the en-US, es-ES and fr-FR locales
- fixed issues:
    - the "Extra menu" action is no longer available in extra menus
    - the pie menu is now closed before an action is run (and it no longer remains open when e.g. the action opens a new window)
- internal improvements mainly in the code managing the pie menu

- This version of easyGestures N deprecates the following items (which might be removed in future versions of the extension):
    - cz-CZ, de-DE, hu-HU, it-IT, ja-JP, ko-KR, pl-PL, pt-BR, sk-SK, tr-TR, zh-CN and zh-TW locales
    - iconless menus

Version 4.8 (December 14, 2014)

- changed actions:
    - "Daily Readings" works now on the bookmarks folder of your choice
    - The 20 "Load URL/Script" actions became 10 "Load URL" actions and 10 "Run Script" actions
    - "Load URL" actions can now open a link in a private window
    - The "Save Link As", "Daily Readings", "Load URL" and "Run Script" actions can now appear as being disabled
- minor improvements in the backend of the preferences dialog


Version 4.7 (October 12, 2014)

- reworked code allowing actions to be disabled
- the following actions can now also appear as being disabled: "Page Top", "Page Bottom", "Close Tab", "Open Link", "Open Link in New Window", "Open Link in New Private Window", "Copy Link", "Bookmark This Link", "Copy Image Location", "Copy Image", "Save Image As", "Hide Images", "Cut", "Copy", "Paste", "Undo", "Select All"
- changed actions:
    - "New Tab", "Close Tab", "New Window" and "Bookmark This Page" mimic now Firefox's behavior
    - "Close Tab" and "Close Other Tabs" take now into account pinned tabs
    - "Duplicate Window" duplicates now all tabs
    - "Open Link", "Open Link in New Window" and "Open Link in New Private Window" only work now on a link
- new actions: "New Blank Tab", "Pin/Unpin Tab", "Close Window", "Undo Close Window", "Show Bookmarks", "Show History", "Show Downloads", "Redo"
- removed action "Mark Visited Links"
- new default menus


Versions 4.6.1 and 4.6.2 (July 18, 2014)

- fixed some warnings issued by AMO scripts


Version 4.6 (July 16, 2014)

- new CSS-based context and alternative menus indicators
- new actions: "New Blank Window" and "Open Link in New Private Window" (suggested by Steve Beckwitt)
- fixed: "Homepage" action could not be called
- refactored code dealing with the execution and management of actions


Version 4.5.3 (May 11, 2014)

- fixed some issues


Version 4.5.2 (April 27, 2014)

- made internal changes to comply with AMO requirements


Version 4.5.1 (April 25, 2014)

- fixed: menu was not properly displayed when moved


Version 4.5 (September 2012--April 2014)

- made easyGestures work with recent versions of Firefox
- made easyGestures a restartless extension
- removed support for upgrading from versions prior to 4.3.2
- moved preferences to the "extensions." branch
- made several changes in the preferences dialog, among which:
    - removed support for multiple skins
    - removed popup menu to change language
    - split the "Behavior" pane in two
    - made use of XUL's preferences system
    - made select boxes for actions non editable
- removed features:
    - input box for entering URLs
    - linear menu
    - red tag for links
    - custom "About" dialog
    - welcome page
    - statusbar icon
    - ability to open menu only on drag
    - "omit tooltips delay" option
    - popup for "Next Tab", "Previous Tab" and "Undo Close Tab" actions
    - "Translate/Explain", "Run Program File", "Mail", "Properties" and "Send Link" actions
- changed features:
    - "Search Web" action uses Firefox's search engines
    - "Highlight/Clear" action toggles Firefox's find bar and was renamed to "Toggle Find Bar"
    - "Daily Readings" action no longer provides default bookmarks upon creation of the daily readings folder and opens all bookmarks in new tabs
    - "Private Browsing" action was changed to "New Private Window"
    - "Close Browser" action was renamed to "Quit Firefox"
- added features:
    - "Restart Firefox" and "Focus Location Bar" actions


chrome/content/tips.js
/***** BEGIN LICENSE BLOCK *****
Version: MPL 1.1/GPL 2.0/LGPL 2.1

The contents of this file are subject to the Mozilla Public License Version 
1.1 (the "License"); you may not use this file except in compliance with 
the License. You may obtain a copy of the License at 
http://www.mozilla.org/MPL/

Software distributed under the License is distributed on an "AS IS" basis,
WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
for the specific language governing rights and limitations under the
License.

The Original Code is easyGestures.

The Initial Developer of the Original Code is Ons Besbes.

Contributor(s): ngdeleito

Alternatively, the contents of this file may be used under the terms of
either the GNU General Public License Version 2 or later (the "GPL"), or
the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
in which case the provisions of the GPL or the LGPL are applicable instead
of those above. If you wish to allow use of your version of this file only
under the terms of either the GPL or the LGPL, and not to allow others to
use your version of this file under the terms of the MPL, indicate your
decision by deleting the provisions above and replace them with the notice
and other provisions required by the GPL or the LGPL. If you do not delete
the provisions above, a recipient may use your version of this file under
the terms of any one of the MPL, the GPL or the LGPL.

***** END LICENSE BLOCK *****/


/* global eGPrefs */

function tipEntry(tipLabelIndex, imageClass, paneName, tabNumber) {
  this.description = tipLabels[tipLabelIndex];
  this.imageClass = imageClass; // image source
  this.paneName = paneName;
  this.tabNumber = tabNumber;
}

var tips = [
  new tipEntry(0,  "triggerBySelection",  "activation",     "0"),
  new tipEntry(1,  "triggerByStroke",     ""          ,     "0"),
  new tipEntry(2,  "alternativeMenus",    "activation",     "0"),
  new tipEntry(3,  "builtInAction",       "menus",          "0"),
  new tipEntry(4,  "clickOnLink",         "",               "0"),
  new tipEntry(5,  "noTooltips",          "behavior",       "0"),
  new tipEntry(6,  "empty",               "",               "0"),
  new tipEntry(7,  "empty",               "",               "0"),
  new tipEntry(8,  "bookmarksOnlyLayout", "customizations", "0"),
  new tipEntry(9,  "programsAndScripts",  "customizations", "1"),
  new tipEntry(10, "contextual",          "activation",     "0"),
  new tipEntry(11, "suppressMenu",        "",               "0"),
  new tipEntry(12, "moveMenu",            "",               "0"),
  new tipEntry(13, "noIcons",             "behavior",       "0"),
  new tipEntry(14, "largeMenu",           "behavior",       "0"),
  new tipEntry(15, "empty",               "behavior",       "0"),
  new tipEntry(16, "empty",               "",               "0"),
  new tipEntry(17, "empty",               "",               "0"),
  new tipEntry(18, "empty",               "customizations", "2")
];

var tipNbr;
document.getElementById("showTips").checked = eGPrefs.areStartupTipsOn();
try {
  tipNbr = eGPrefs.getTipNumberPref();
}
catch (ex) {
  tipNbr = 1;
}

function updateTipNbr(step) {
  tipNbr = (((tipNbr + step) % tips.length) + tips.length) % tips.length;
  eGPrefs.setTipNumberPref(tipNbr);
  updateContent(tipNbr);
}

function showPaneAndTabOn(paneName, tabNumber, doc) {
  var pane = doc.getElementById(paneName + "Pane");
  doc.getElementById("eG_optionsWindow").showPane(pane);
  if (tabNumber > 0) {
    var tabbox = doc.getElementById(paneName + "Tabboxes");
    tabbox.selectedIndex = tabNumber;
  }
}

function openPreferencesWith(paneName, tabNumber) {
  var winenum = Services.wm.getEnumerator("");
  var found = false;
  var win;
  while (winenum.hasMoreElements() && !found) {
    win = winenum.getNext();
    found = win.location == "chrome://easygestures/content/options.xul";
  }
  if (!found) {
    win = window.openDialog("chrome://easygestures/content/options.xul", "", "");
    win.addEventListener("load", function runOnLoad() {
      removeEventListener("load", runOnLoad, false);
      showPaneAndTabOn(paneName, tabNumber, win.document);
    }, false);
  }
  else {
    showPaneAndTabOn(paneName, tabNumber, win.document);
    win.focus();
  }
}

function tipLinkClick() {
  var paneName = tips[tipNbr].paneName;
  var tabNumber = tips[tipNbr].tabNumber;
  openPreferencesWith(paneName, tabNumber);
}

function updateContent(tipNbr) {
  document.getElementById("tipNbrDisplay").value = (tipNbr + 1) + " / " + tips.length;
  
  var text = tips[tipNbr].description.split("|")[0];
  var linkText = (tips[tipNbr].description.split("|").length > 1 ?
                    tips[tipNbr].description.split("|")[1]
                  : "");
  
  document.getElementById("tipText").textContent = text;
  document.getElementById("tipImage").setAttribute("class", tips[tipNbr].imageClass);
  document.getElementById("tipLink").textContent = linkText;
  sizeToContent();
}

function copyDescriptionToClipboard() {
  var description = tips[tipNbr].description.split("|")[0];
  Components.classes["@mozilla.org/widget/clipboardhelper;1"].getService(Components.interfaces.nsIClipboardHelper).copyString(description);
}